docs/rfd.md

```diff
@@ -0,0 +1,478 @@
+2. When a client initiates a stream by calling `Output()`, we will write the current contents of the buffer to the stream in 4MB chunks. Then, if the client requested to follow the output, the server will keep the stream open, and any subsequent reads from the process's output will be duplicated to all clients that are following the stream, in addition to the server's own buffer.
+
+#### Process Lifecycle
+
+For jobs that do not specify resource limits, they will be run as a child process of the job server itself without any additional isolation. For jobs that do specify resource limits, they will be run in their own cgroup with all of the requested limits.
+
+On startup, the server will do the following (simplifying, there will be additional abstractions here):
+
+1. Ensure its euid is 0 so it can manage cgroups, and ensure that cgroups v2 is enabled.
+2. Create a parent cgroup in `/user.slice/user-<uid>.slice/jobserver` if it doesn't exist. The specific path can be determined by first identifying its own cgroup from `/proc/self/cgroup` (which, if run as `sudo jobserver serve`, will be something like `/user.slice/user-<uid>.slice/<terminal emulator's cgroup>/vte-spawn-<some uuid>.scope`), then walking the tree up until it finds a cgroup matching `user-*.slice`.
+3. Enable the `cpu`, `memory`, and `io` controllers in the `jobserver` cgroup's `cgroup.subtree_control` file, if necessary.
+
+The lifecycle of a job's cgroup will be as follows:
+
+1. When a job is created with resource limits, the server will create a new cgroup with the name `<job-id>` under the parent `jobserver` cgroup.
+2. The server will write the requested resource limits to the new cgroup's `cpu.max`, `memory.max`, and `io.max` files (for whichever limits were specified).
+3. The server will create a file descriptor to the new cgroup by opening its path (`/sys/fs/cgroup/user.slice/user-<uid>.slice/jobserver/<job-id>`).
+4. The server will configure the `exec.Cmd`'s `SysProcAttr` with `CgroupFd` set to the file descriptor. When this is set, Go will call `clone3` with `CLONE_INTO_CGROUP` so that the new child process is started directly in the new cgroup.
+5. When the process exits, the server will close the file descriptor and delete the job's cgroup.
+
+Other notes:
+
+- When clients are streaming logs for a running process, the server will close those streams once the process has exited and all logs have been written.
+- The server will not attempt to modify the controllers of its parent cgroups, so if `user-<uid>.slice` doesn't have any of the required controlers enabled, it will simply bail out.
+- The user will be able to override the default parent cgroup path with a flag, since there can be some variation in the cgroup hierarchy depending on the system.
+
+### Security
+
+#### Authentication and Authorization
+
+Authentication is implemented using simple mTLS, with a subject name encoded in the client certificate's common name field.
```

Comment:
Nit: Can you specify what exactly will be encoded in the subject? I think it's _pretty_ clear that it will be the "subject" from the authz map but I didn't see it explicitly mentioned anywhere so wanted to make sure my understanding is correct.

Reply:
That is correct, the client certificate's Common Name is used as the subject name for authz purposes. So the user "admin" will authenticate with a client cert that looks like:
![image](https://github.com/kralicky/jobserver/assets/8194899/f787a35a-d604-4e02-8585-724a7fa3369a)

Reply:
Let me know if that answers https://github.com/kralicky/jobserver/pull/1#discussion_r1424672671 as well.

Reply:
I'd use a different word since the "subject" in X.509 is the whole RDNSequence of fields, not just the string in the CN. Maybe "user" or "username"?